### ISSUE TITLE
[PyTorch] proposed enhancement: make varargs more flexible

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#7674]
We currently only support int varargs when they are the only argument in the Python torch wrapping. This limits, for example, the use of varargs in einsum.

The root cause is in [the pattern for using `PythonArgParser` with the templated `ParsedArgs` as documented in the comment at the top of `torch/csrc/utils/python_arg_parser.h`](https://github.com/pytorch/pytorch/blob/c58893eb9eb9fca2c61aaee1c7013983569dad61/torch/csrc/utils/python_arg_parser.h#L14).
The fact that `ParsedArgs` are accessible without `PythonArgParser` keeps us from doing Python object allocations (for a slice of arguments to assign to the vararg) that could be cleaned up in the destructor of `PythonArgParser` without feeling dirty.
I see two obivous ways out:
1. move the `ParsedArgs` structure into PythonArgParser and make that templated (but I would guess  that there was a reason to not do that all along - possibly duplicated functions)
2. add a `std::vector<PyObject*> varargs` to `ParsedArgs` that gets the varargs.

Would either of these be acceptable?

Best regards

Thomas

**Comment by zou3519** [comment]
How would you use varargs in einsum? Would it look something like (borrowing from the docs): `torch.einsum('bn,anm,bm->ba', l,A,r)`?

**Comment by t-vi** [comment]
Exactly. That is what numpy does and I keep forgetting it even though I wrote both the function and the docs. If einsum really is the only user, it might not be worth it, but I might just wrap einsum manually, but if there are more "flexible number of tensor args" or more than just varargs cases, they might benefit, too.

**Comment by apaszke** [comment]
I think that since einsum is the only case where we could possibly need this (at least for now), then we should just have a special case in our code (perhaps add a new type to PythonArgParser, that would be handled specially). Making all other places pay the complexity for a quite exotic case is not good.
